pkg/controller/shared_state.go

```diff
@@ -555,8 +556,80 @@ func (c *SharedState) addVolumeAndClaim(azVolumeName, pvName, pvClaimName string
 	c.claimToVolumeMap.Store(pvClaimName, azVolumeName)
 }
 
-func (c *SharedState) deletePV(pvName string) {
+func (c *SharedState) deletePV(pvName string) error {
 	c.pvToVolumeMap.Delete(pvName)
+
+	// fetch AzVolume CRI for the PV, and delete if it's a preprovisioned volume
+	ctx := context.Background()
+	var azVolumeList azdiskv1beta2.AzVolumeList
+	req, err := azureutils.CreateLabelRequirements(consts.PvNameLabel, selection.Equals, pvName)
+	if err != nil {
+		return err
+	}
+	err = c.cachedClient.List(ctx, &azVolumeList, &client.ListOptions{LabelSelector: labels.NewSelector().Add(*req)})
+	if err != nil && !apiErrors.IsNotFound(err) {
+		return err
+	} else if apiErrors.IsNotFound(err) || len(azVolumeList.Items) == 0 {
+		return nil
+	}
+
+	// deletion timestamp is set and AzVolume reconcliler will handle the delete request.
+	// The volume itself will not be deleted.
+	azVolume := azVolumeList.Items[0]
+
+	if azureutils.MapContains(azVolume.Status.Annotations, consts.PreProvisionedVolumeAnnotation) {
+		return nil
+	}
+
+	// AzVolume does exist and needs to be deleted
+	// add annotation to mark AzVolumeAttachment cleanup
+	err = c.cachedClient.Delete(ctx, &azVolume)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	klog.Info("azVolume deleted")
+
+	azVolumeNamespacedName := types.NamespacedName{Namespace: azVolume.Namespace, Name: azVolume.Name}
+	azVolume = *azVolume.DeepCopy()
+	updateFunc := func(obj client.Object) error {
+		azVolume := obj.(*azdiskv1beta2.AzVolume)
+		azVolume.Status.Annotations = azureutils.AddToMap(azVolume.Status.Annotations, consts.PreProvisionedVolumeCleanupAnnotation)
+		return nil
+	}
+	_, err = azureutils.UpdateCRIWithRetry(ctx, nil, c.cachedClient, c.azClient, &azVolume, updateFunc, consts.ForcedUpdateMaxNetRetry, azureutils.UpdateCRIStatus)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	klog.Info("azVolume updated")
+
+	go func() {
+		_ = wait.PollImmediateInfinite(consts.CRIUpdateRetryDuration, func() (bool, error) {
```

Comment:
Can we use a condition waiter instead of a polling wait? Polling impacts API server throttling.